src/py/flwr/common/auth_plugin/auth_plugin.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Abstract classes for Flower User Auth Plugin."""
+
+
+from abc import ABC, abstractmethod
+from collections.abc import Sequence
+from pathlib import Path
+from typing import Any, Optional, Union
+
+from flwr.proto.exec_pb2 import (  # pylint: disable=E0611
+    GetAuthTokensRequest,
+    GetAuthTokensResponse,
+    GetLoginDetailsResponse,
+)
+from flwr.proto.exec_pb2_grpc import ExecStub
+
+
+class ExecAuthPlugin(ABC):
+    """Abstract Flower Auth Plugin class for ExecServicer."""
+
+    @abstractmethod
+    def __init__(self, config: dict[str, Any]):
+        """Abstract constructor (init)."""
+
+    @abstractmethod
+    def get_login_details(self) -> GetLoginDetailsResponse:
+        """Get the GetLoginDetailsResponse containing the login details."""
+
+    @abstractmethod
+    def validate_tokens_in_metadata(
+        self, metadata: Sequence[tuple[str, Union[str, bytes]]]
+    ) -> bool:
+        """Validate the auth tokens in the provided metadata."""
+
+    @abstractmethod
+    def get_auth_tokens(self, request: GetAuthTokensRequest) -> GetAuthTokensResponse:
+        """Get the relevant auth tokens."""
+
+    @abstractmethod
+    def refresh_tokens(
+        self, metadata: Sequence[tuple[str, Union[str, bytes]]]
+    ) -> Optional[Sequence[tuple[str, Union[str, bytes]]]]:
+        """Refresh auth tokens in the provided metadata."""
+
+
+class CliAuthPlugin(ABC):
+    """Abstract Flower Auth Plugin class for CLI."""
+
+    @staticmethod
+    @abstractmethod
+    def login(
+        login_details: dict[str, str],
+        config: dict[str, Any],
```

Comment:
Wouldn't it better to make use of some of the types we have for configs ? For example https://github.com/adap/flower/blob/fc8591fce970c2a533deb0000427d24075797205/src/py/flwr/common/typing.py#L65

Reply:
This function expects a json dictionary here, and it would be rather flexible if we just use `dict[str, Any]`. Though we normally just need to store scalars in the json dict, I am not quite sure what's the benefit of enforcing a stricter type here.

Reply:
It receives a `{string: string}` (https://github.com/adap/flower/blob/0730b12a9e9653782c7ee1777c4382f50380217a/src/proto/flwr/proto/exec.proto#L67) unless we modify the `exec.proto`. Happy to leave it as `Any`, but in other parts of the code we are more string with these things (e.g. `typing.UserConfig`)